### PR TITLE
[RDY] Fix build error with Microsoft Visual Studio

### LINKED ISSUE #16399: MSVC140 Project does not compile
There are at least a few issues with MSVC compilation.

1. MSVC as a compiler does not support C99 compound literals. Input.cpp(782) has one.
This can be fixed fairly easily using different ISO C++ standard techniques.

2. ..\src\explosion.cpp(52): fatal error C1001: An internal error has occurred in the compiler. 
I don't know enough about C++ or MSVC to diagnose this issue.

3. Visual Studios 2015 Update 2 utilizes new Intellisense cache files that should be added to .gitignore
msvc140\Cataclysm.VC.db
msvc140\Cataclysm.VC.VC.opendb
msvc140\ipch\*

I can fairly easily fix 1 & 3, and submit a PR, but I have no idea what's going on with explosion.cpp(52). I've tried google and MSDN, but it's just beyond my C++ skillset. Until that's resolved I can't verify any MSVC compiling, so I'd rather not submit a PR with a "Partial fix"

Edit: Self-Reproducible on fresh machine with a fresh install of Visual Studio Community 2015 Update 2

### PR DESCRIPTION
**Pull request by ghost**
This partially fixes #16399 and #12426. After these changes, Visual Studio 2015 with Update 2 is able to build CDDA successfully. Tiles and localization are both functional.

### CODE REVIEW

File: src/output.cpp [hunk 1]
```diff
@@ -174,7 +174,7 @@ void trim_and_print(WINDOW *w, int begin_y, int begin_x, int width, nc_color bas
             iLength += iTempLen;
 
             if ( iLength > width ) {
-                sTempText = sTempText.substr(0, cursorx_to_position(sTempText.c_str(), iTempLen - (iLength - width) - 1, NULL, -1)) + "…";
+                sTempText = sTempText.substr(0, cursorx_to_position(sTempText.c_str(), iTempLen - (iLength - width) - 1, NULL, -1)) + "...";
```
Comment (illi-kun) [inline src/output.cpp]: This is a bad replacement. It steal two more meaning symbols from text.

File: src/explosion.cpp [hunk 2]
```diff
@@ -50,6 +50,7 @@ shrapnel_data load_shrapnel_data( JsonObject &jo )
     return ret;
 }
 
+// (C1001) Comiler Internal Error on Visual Studio 2015 with Update 2
```
Comment (illi-kun) [inline src/explosion.cpp]: Minor note: Typo in 'Com_p_iler' spelling.

File: src/output.h [hunk 3]
```diff
@@ -427,9 +428,39 @@ std::pair<std::string, nc_color> const& get_light_level(const float light);
...
+            }
+            used_width += seg_width;
+            result.insert(result.end(), seg_width, it->second);
+        }
+        result.insert(result.end(), bar_width - used_width, ' ');
+    }
+    result += ']';
+
+    return result;
+}
```
Comment (codemime) [inline src/output.h]: Why did you move the whole function to the header?

Comment (codemime) [inline src/output.h]: Please follow [the coding style](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.txt). The function was astyled correctly prior to your changes.

File: src/output.cpp (hunk 1 above)
Comment (DLaboratory) [inline src/output.cpp]: @illi-kun Thank you reading my changes. When using the single length character "…", Visual Studio gives a compile error.

File: src/output.h (hunk 3 above)
Comment (DLaboratory) [inline src/output.h]: @codemime MSVC seems has problem dealing with template functions. Link error appears if the implementation of function `get_labeled_bar` is in `output.cpp`. After I move the function implementation to header file, it passes the compilation.

File: src/output.cpp (hunk 1 above)
Comment (illi-kun) [inline src/output.cpp]: No problem. I understand why you change that but I disagree with solution. Could you try to use "\u2026" sequence (with comment what is it).

BTW, this solution is used in function ```utf8_wrapper::shorter()``` in ```src\catacharset.cpp``` (line ~602, close to the end of file), so you can use it as example, or even modify functions for deduplication.

**Comment by DLaboratory** [comment]
This PR is done.

File: src/bionics_ui.cpp [hunk 4]
```diff
@@ -73,7 +73,7 @@ void draw_bionics_titlebar( WINDOW *window, player *p, bionic_menu_mode mode )
 
 const auto separator = []( std::ostringstream &s )
 {
-    return s.tellp() != 0 ? ", " : "";
+    return (int)(s.tellp()) != 0 ? ", " : "";
```
Comment (illi-kun) [inline src/bionics_ui.cpp]: Style regression.

It should looks like this: ```return ( int )( s.tellp() ) != 0 ? ", " : "";``` (please pay your attention to whitespaces).

Comment (DLaboratory) [inline src/bionics_ui.cpp]: @illi-kun May I just run `astyle` for those 8 files I changed?

**Comment by DLaboratory** [comment]
```
ccache clang++  -DRELEASE -DGIT_VERSION -DLOCALIZE -ffast-math -Os  -Wall -Wextra   -Werror -std=c++11 -MMD -I/usr/include/lua5.2   -DLUA -I/usr/include/ncursesw -c src/wish.cpp -o obj/wish.o
ccache clang++  -DRELEASE -DGIT_VERSION -DLOCALIZE -ffast-math -Os  -Wall -Wextra   -Werror -std=c++11 -MMD -I/usr/include/lua5.2   -DLUA -I/usr/include/ncursesw -c src/worldfactory.cpp -o obj/worldfactory.o
lang/compile_mo.sh cs de el es_AR es_ES fi fr it_IT ja ko pl pt_BR pt pt_PT ru sr vi zh_CN zh_TW
astyle regressions found.
Formatted  src/item_factory.h
make: *** [astyle-check] Error 1
make: *** Waiting for unfinished jobs....
make: *** wait: No child processes.  Stop.
Build step 'Execute shell' marked build as failure
[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] Skipped based on build state FAILURE
Finished: FAILURE
```

Does it mean I need to `astyle` all source code to pass the check?

**Comment by DLaboratory** [comment]
oops, I'd better revert the astyle commit

**Comment by DLaboratory** [comment]
Is there any problem before this PR can be merged?

**Comment by BorkBorkGoesTheCode** [comment]
You may want to look at #16928

**Comment by BorkBorkGoesTheCode** [comment]
All the PRs are affected.

**Comment by DLaboratory** [comment]
Aha, all checks have passed.

**Comment by DLaboratory** [comment]
Trying to build a 64 bit gettext library from source with MinGW w64, but strange `windres` make error appears.... Confused....

**Comment by DLaboratory** [comment]
I give up adding 64-bit libintl support. It seems a little difficult. I'll do that in the future, not in this PR. Now this PR is ready to be merged.

**Pull request merged**